File: src/kimi_cli/web/api/open_in.py
"""Open local apps for a path on the host machine."""

from __future__ import annotations

import shlex
import subprocess
import sys
from pathlib import Path
from typing import Literal

from fastapi import APIRouter, HTTPException, status
from loguru import logger
from pydantic import BaseModel

router = APIRouter(prefix="/api/open-in", tags=["open-in"])


class OpenInRequest(BaseModel):
    """Open path in a local app."""

    app: Literal["finder", "cursor", "vscode", "iterm", "terminal"]
    path: str


class OpenInResponse(BaseModel):
    """Open path response."""

    ok: bool
    detail: str | None = None


def _resolve_directory(path: str) -> Path:
    resolved = Path(path).expanduser()
    try:
        resolved = resolved.resolve()
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Path does not exist: {path}",
        ) from None

    if not resolved.exists():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Path does not exist: {path}",
        )
    if not resolved.is_dir():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Path is not a directory: {path}",
        )
    return resolved


def _run_command(args: list[str]) -> None:
    subprocess.run(
        args,
        check=True,
        capture_output=True,
        text=True,
    )


def _open_app(app_name: str, path: Path, fallback: str | None = None) -> None:
    try:
        _run_command(["open", "-a", app_name, str(path)])
        return
    except subprocess.CalledProcessError as exc:
        if fallback is None:
            raise
        logger.warning("Open with {} failed: {}", app_name, exc)
    _run_command(["open", "-a", fallback, str(path)])


def _open_terminal(path: Path) -> None:
    command = f"cd {shlex.quote(str(path))}"
    script = f'tell application "Terminal" to do script "{command}"'
    _run_command(["osascript", "-e", script])


def _open_iterm(path: Path) -> None:
    command = f"cd {shlex.quote(str(path))}"
    script = "\n".join(
        [
            'tell application "iTerm"',
            "  create window with default profile",
            "  tell current session of current window",
            f'    write text "{command}"',
            "  end tell",
            "end tell",
        ]
    )
    try:
        _run_command(["osascript", "-e", script])
    except subprocess.CalledProcessError:
        script = script.replace('"iTerm"', '"iTerm2"')
        _run_command(["osascript", "-e", script])


@router.post("", summary="Open a path in a local application")
async def open_in(request: OpenInRequest) -> OpenInResponse:
    if sys.platform != "darwin":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Open-in is only supported on macOS.",
        )

    directory = _resolve_directory(request.path)

    try:
        match request.app:
            case "finder":
                _run_command(["open", str(directory)])
            case "cursor":
                _open_app("Cursor", directory)
            case "vscode":
                _open_app("Visual Studio Code", directory, fallback="Code")
            case "iterm":
                _open_iterm(directory)
            case "terminal":
                _open_terminal(directory)
            case _:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Unsupported app: {request.app}",
                )
    except subprocess.CalledProcessError as exc:
        logger.warning("Open-in failed ({}): {}", request.app, exc)
        detail = exc.stderr.strip() if exc.stderr else "Failed to open application."
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=detail,
        ) from exc

    return OpenInResponse(ok=True)

File: src/kimi_cli_ts/web/api/open_in.ts
/**
 * Web API open-in — corresponds to Python web/api/open_in.py
 * Open files/folders in local applications (macOS, Windows).
 * Nearly identical to vis/app.ts handleOpenIn, but as a standalone handler.
 */

import { existsSync, statSync } from "node:fs";
import { resolve } from "node:path";

// ── Types ────────────────────────────────────────────────

interface OpenInRequest {
	app: string;
	path: string;
}

const VALID_APPS = [
	"finder",
	"cursor",
	"vscode",
	"iterm",
	"terminal",
	"antigravity",
];

// ── Helpers ──────────────────────────────────────────────

function jsonResponse(data: unknown, status = 200): Response {
	return new Response(JSON.stringify(data), {
		status,
		headers: {
			"Content-Type": "application/json; charset=utf-8",
			"Access-Control-Allow-Origin": "*",
		},
	});
}

// ── Platform-specific openers ────────────────────────────

async function openInMacOS(
	app: string,
	path: string,
	isFile: boolean,
): Promise<void> {
	const proc = (args: string[]) =>
		Bun.spawn(args, { stdout: "pipe", stderr: "pipe" });

	switch (app) {
		case "finder":
			if (isFile) {
				await proc(["open", "-R", path]).exited;
			} else {
				await proc(["open", path]).exited;
			}
			break;
		case "cursor":
			await proc(["open", "-a", "Cursor", path]).exited;
			break;
		case "vscode":
			try {
				await proc(["open", "-a", "Visual Studio Code", path]).exited;
			} catch {
				await proc(["open", "-a", "Code", path]).exited;
			}
			break;
		case "antigravity":
			await proc(["open", "-a", "Antigravity", path]).exited;
			break;
		case "iterm": {
			const dir = isFile ? resolve(path, "..") : path;
			const script = [
				'tell application "iTerm"',
				"  create window with default profile",
				"  tell current session of current window",
				`    write text "cd " & quoted form of "${dir}"`,
				"  end tell",
				"end tell",
			].join("\n");
			try {
				await proc(["osascript", "-e", script]).exited;
			} catch {
				await proc(["osascript", "-e", script.replace('"iTerm"', '"iTerm2"')])
					.exited;
			}
			break;
		}
		case "terminal": {
			const dir = isFile ? resolve(path, "..") : path;
			const script = `tell application "Terminal" to do script "cd " & quoted form of "${dir}"`;
			await proc(["osascript", "-e", script]).exited;
			break;
		}
	}
}

async function openInWindows(
	app: string,
	path: string,
	isFile: boolean,
): Promise<void> {
	const proc = (args: string[]) =>
		Bun.spawn(args, { stdout: "pipe", stderr: "pipe" });

	switch (app) {
		case "finder":
			if (isFile) {
				Bun.spawn(["explorer", `/select,${path}`]);
			} else {
				Bun.spawn(["explorer", path]);
			}
			break;
		case "cursor":
			await proc(["cmd", "/c", "start", "", "cursor", path]).exited;
			break;
		case "vscode":
			await proc(["cmd", "/c", "start", "", "code", path]).exited;
			break;
		case "terminal": {
			const dir = isFile ? resolve(path, "..") : path;
			try {
				await proc(["cmd", "/c", "start", "", "wt.exe", "-d", dir]).exited;
			} catch {
				await proc([
					"cmd",
					"/c",
					"start",
					"",
					"cmd.exe",
					"/K",
					`cd /d "${dir}"`,
				]).exited;
			}
			break;
		}
		case "iterm":
		case "antigravity":
			throw new Error(`${app} is not supported on Windows.`);
	}
}

// ── Route handler ────────────────────────────────────────

export async function handleOpenInRoute(req: Request): Promise<Response> {
	const { platform } = process;
	if (platform !== "darwin" && platform !== "win32") {
		return jsonResponse(
			{ detail: "Open-in is only supported on macOS and Windows." },
			400,
		);
	}

	let body: OpenInRequest;
	try {
		body = (await req.json()) as OpenInRequest;
	} catch {
		return jsonResponse({ detail: "Invalid JSON body" }, 400);
	}

	if (!VALID_APPS.includes(body.app)) {
		return jsonResponse({ detail: `Unsupported app: ${body.app}` }, 400);
	}

	const resolvedPath = resolve(body.path);
	if (!existsSync(resolvedPath)) {
		return jsonResponse({ detail: `Path does not exist: ${body.path}` }, 400);
	}

	const isFile = statSync(resolvedPath).isFile();

	try {
		if (platform === "darwin") {
			await openInMacOS(body.app, resolvedPath, isFile);
		} else {
			await openInWindows(body.app, resolvedPath, isFile);
		}
	} catch (err) {
		const detail =
			err instanceof Error ? err.message : "Failed to open application.";
		return jsonResponse({ detail }, 500);
	}

	return jsonResponse({ ok: true });
}
